quil-rs/src/validation/identifier.rs

```diff
@@ -0,0 +1,34 @@
+//! Types and functions related to validating Quil identifiers
+use std::str::FromStr;
+
+use fancy_regex::Regex;
+use thiserror;
+
+use crate::reserved::ReservedToken;
+
+#[derive(Debug, PartialEq, thiserror::Error)]
+pub enum IdentifierValidationError {
+    #[error("{0} is a reserved token")]
+    Reserved(ReservedToken),
+
+    #[error("{0} is not a valid identifier")]
+    Invalid(String),
+}
+
+/// A regex that matches only valid Quil identifiers
+const IDENTIFIER_REGEX: &str = r"^[A-Za-z_]+[A-Za-z0-9\-_]*.*(?<!\-)$";
```

Comment:
The `.*` seems incorrect to me. Why not just use the pattern directly from the spec? `[A-Za-z_][A-Za-z0-9\-_]*[A-Za-z0-9_]`

Reply:
I started with that initially, but it doesn't work for 1 character strings like `H`. Attempting to patch that causes other problems with the regex.

The `quil` spec actually defines _two_ regexes for an identifier which didn't strike me at the time as something that could be composed into a single regex, but of course it can, so I stitched them together and switched back to the `regex` crate.

I also pushed up some tests for validation.